Add functions to modify element collections

CameronSWilliamson opened the pull request.

CameronSWilliamson committed:
Added Get-CIisCollection and Get-CIisCollectionKey

CameronSWilliamson committed:
Added Set-CIisCollection function

CameronSWilliamson committed:
Set-CIIsCollection saves only on change

CameronSWilliamson committed:
Added Add-CIisCollectionItem function

CameronSWilliamson committed:
Added Strict to functions that didn't have it

CameronSWilliamson committed:
Created Add-CIisHttpHeader function

CameronSWilliamson committed:
Added documentation and cleaned up variable names

CameronSWilliamson committed:
Added Remove-CIisCollectionitem function

CameronSWilliamson committed:
Incrementing version

splatteredbits commented on CHANGELOG.md:
```diff
@@ -3,6 +3,17 @@
+
+## 1.1.0
+
+Added: `Get-CIisCollection` for accessing items in a specific IIS collection.
```
We try to follow the [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) format.

* Move the `> Released` section under the version header.
* Fix the release date, since I took so long to get to this.
* Instead of duplicating `Added:` on every line, add a `### Added` heading, then remove `Added: ` with *what* was added.
* When changing multiple things, the changelog should be a list. Add a `*` to each item.

For example,

```markdown
## 1.1.0

> Released 16 Jun 2023

### Added

* Function `Get-CIisCollection` for accessing items in a specific IIS collection.
* Function `Get-CIisCollectionKey` for determining unique key for an IIS collection.
* Function `Set-CIisCollection` for setting an IIS collection's elements.
* Function `Add-CIisCollectionItem` for adding a new item to an IIS collection.
* Function `Add-CIisHttpHeader` for adding custom HTTP headers to IIS.
* Function `Remove-CIisCollectionItem` for removing an item from an IIS collection.
```

splatteredbits commented on Carbon.IIS/Functions/Add-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,155 @@
+
+    #>
+    [CmdletBinding(DefaultParameterSetName='GlobalBySingleKey')]
+    param(
```
Missing parameter documentation.

splatteredbits commented on Carbon.IIS/Functions/Add-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,155 @@
+        [Parameter(Mandatory, ParameterSetName='LocationBySingleKey')]
+        [Parameter(Mandatory, ParameterSetName='GlobalBySingleKey')]
+        [AllowNull()]
+        [Object] $Value,
```
Why are you allowing `$null`? Not sure it makes sense to allow/have an item's key attribute be null.

splatteredbits commented on Carbon.IIS/Functions/Add-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,155 @@
+        $collectionArgs['Name'] = $CollectionName
+    }
+
+    $collection= Get-CIisCollection @collectionArgs
```
Lint: space before `=`.

splatteredbits commented on Carbon.IIS/Functions/Add-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,155 @@
+    .EXAMPLE
+    Add-CIisCollectionItem -LocationPath 'SITE_NAME' -SectionPath 'system.webServer/defaultDocument' -CollectionName 'files' -Value 'hello.htm'
+
+    Demonstrates how to add an item in an IIS collection with the provided value.
```
Rephrase to "Demonstrates how to add an item to a named collection under a configuration section. After the above command runs, this will be added to the applicationHost.config:"

Lint: break lines at 120, not at the end of a sentence.

splatteredbits commented on Carbon.IIS/Functions/Add-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,155 @@
+{
+    <#
+    .SYNOPSIS
+    Adds a new `Microsoft.Web.Administration.ConfigurationElement` with the add element name to a collection.
```
Docs should be a little less technical and a little higher-level.

```text
.SYNOPSIS
Adds a new item to an IIS configuration collection.

.DESCRIPTION
The `Add-CIisCollectionItem` function adds a new item to an IIS configuration collection. Pass the collection's IIS configuration section path to the `SectionPath` parameter and the value to add to the collection to the `Value` parameter. The function adds that value to the collection if it doesn't already exist in the collection.

To add extra attributes to the collection item, pass the *extra* attributes as a hashtable to the `Attribute` parameter. If the item already exists in the collection, any attributes on the item that aren't in the `Attribute` hasthable are *removed* from the collection item.

If the configuration section given by `SectionPath` isn't a collection, pass the name of the collection to the `CollectionName` parameter.

If adding an item to the collection for a website, application, virtual directory, or directory, pass that path to that location to the `LocationPath` parameter.
```
*

splatteredbits commented on Carbon.IIS/Functions/Add-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,155 @@
+    This function can also add additional fields using the `$Attribute` paramater as long as the unique key is provided in the hashtable.
+    After adding an item, the changes to applicationHost.config are committed and saved automatically.
+
+    .EXAMPLE
```
Example should go from simplest to most complex, and demonstrate each parameter set and optional parameter. I usually group optional parameters into a single example. The first example should be the simplest, using defaults for all but the mandatory parameters. Then add a second example with all the optional parameters. Then move on to more complex parameter sets.

```text
.EXAMPLE
> Add-CIisCollectionItem -SectionPath 'system.webServer/someCollection' -Value 'someItem'

Demonstrates how to add an item to a collection. In this example, if `someCollection` were a configuration section *and* a collection `<add value="someItem" />` would be added to the `system.webServer/someCollection` collection. This example isn't more specific because we don't know of a configuration section that is itself a collection that only requires a value and not other attributes in the collection item.

.EXAMPLE
> Add-CIisCollectionItem -SectionPath 'system.webServer/globalModules' -Value 'SomeModule' -Attribute @{ image = 'C:\path\to\somemodule.dll" }

Demonstrates how to add additional attributes to a collection item by passing those attributes to the `Attribute` parameter. In this example, will add a `<add name="SomeModule" image="C:\path\to\somemodule.dll" />` item to the `system.webServer/globalModules` collection.

.EXAMPLE
> Add-CIisCollectionItem -SectionPath 'system.webServer/defaultDocument' -CollectionName 'files' -Value 'home.html'

Demonstrates how to add an item into a configuration section's collection. In this example, `<add value="home.html" />` would be added to the `system.webServer/defaultDocument` section's `<files>` collection.

.EXAMPLE
> Add-CIisCollectionItem -LocationPath 'website/directory' -SectionPath 'system.webServer/defaultDocument' -CollectionName 'files' -Value 'hello.htm'

Demonstrates how to add an item to a configuration section's collection but only under a specific website, directory, application, or virtual directory. After running this example, this would be added to IIS's applicationHost.config:

        <location path="website/directory">
            <system.webServer>
                <defaultDocument>
                    <files>
                        <add value="hello.htm" />
                    </files>
                </defaultDocument>
            </system.webServer>
        </location>
```

splatteredbits commented on Carbon.IIS/Functions/Add-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,155 @@
+
+    $uniqueKeyAttrName = Get-CIisCollectionKeyName -Collection $collection
+
+    if (-not $Value)
```
* Reduce code duplication.
* Improve messaging.
* The `Value` parameter should be required and should *not* be in the `Attribute` hashtable (but only warn if it is).
* Remove attributes if they don't exist in the `Attribute` hashtable.

```powershell
    $msgs = [Collections.List[String]]:New()
    $msgs.Add("${SectionPath}$($CollectionName ? "/${CollectionName}" : '')
    
    $keyAttrName = Get-CIisCollectionKeyName -Collection $collection

    $add = $collection | Where-Object { $_.GetAttributeValue($keyAttrName) -eq $Value }

    # Create the add element once, if it doesn't exist.
    if (-not $add)
    {
        $add = $collection.CreateElement('add')
        $add.SetAttributeValue($keyAttrName, $Value)
        [void]$collection.Add($add)
        $msgs.Add("  + ${keyAttrName} = ${Value}")       
        $save = $true
    }
    else
    {
        $msgs.Add("    ${keyAttrName} = ${Value}")
    }
        
    if (-not $Attribute)
    {
        $Attribute = @{}
    }

    foreach ($attrName in $Attribute.Keys)
    {
        if ($attrName -eq $keyAttrName)
        {
            # TODO: create a better warning message
            Write-Warning 'pass as Value parameter, not as an item in Attribute. Attribute is only for extra attributes.'
            continue
        }
        
        $expectedValue = $Attribute[$attrName]
        $actualAttr = $add.GetAttribute($attrName)
        if ($null -eq $actualValue)
        {
            $msgs.Add("      + ${attrName} = ${expectedValue}")
            $add.SetAttributeValue($attrName, $expectedValue)
            $save = $true
            continue
        }
        
        if ($expectedValue -ne $actualAttr.Value)
        {
            $msgs.Add("      ${attrName}  $($actualAttr.Value) → ${expectedValue}")
            $add.SetAttributeValue($attrName, $expectedValue)
            $save = $true
        }
    }
    
    # Remove any extra attributes.
    foreach ($attr in $add)
    {
        if ($attr.Name -eq $keyAttrName)
        {
            continue
        }
        
        if (-not $Attribute.ContainsKey($attr.Name))
        {
            $msgs.Add("    - $($attr.Name)")
            $add.Remove($attr.Name)
            $save = $true
        }
    }

    if ($save)
    {
        foreach ($msg in $msgs)
        {
            Write-Information $msg
        }
        
        Save-CIisConfiguration
    }
```

splatteredbits commented on Carbon.IIS/Functions/Add-CIisHttpHeader.ps1:
```diff
@@ -0,0 +1,72 @@
+function Add-CIisHttpHeader
+{
+    <#
+    .SYNOPSIS
```
Rewrite docs following pattern in comments on the `Add-CIisCollectionItem` function. These are a little too technical.

splatteredbits commented on Carbon.IIS/Functions/Add-CIisHttpHeader.ps1:
```diff
@@ -0,0 +1,72 @@
+        $addParameters['LocationPath'] = $LocationPath
+    }
+
+    Add-CIisCollectionItem @addParameters -Attribute $attributes
```
Only put *optional* parameters in splatted hashtables.

```powershell
Add-CIisCollectionItem -SectionPath $SectionPath -CollectionName 'customHeaders' -Value $Name -Attribute @{ value = $Value }
```

splatteredbits commented on Carbon.IIS/Functions/Get-CIisCollection.ps1:
```diff
@@ -0,0 +1,50 @@
+function Get-CIisCollection
+{
+    <#
```
Ditto on these docs.

splatteredbits commented on Carbon.IIS/Functions/Get-CIisCollection.ps1:
```diff
@@ -0,0 +1,50 @@
+    Use-CallerPreference -Cmdlet $PSCmdlet -Session $ExecutionContext.SessionState
+
+    $sectionArgs = @{
+        "SectionPath" = $SectionPath
```
Lint: use single quotes if there is no interpolation.

splatteredbits commented on Carbon.IIS/Functions/Get-CIisCollection.ps1:
```diff
@@ -0,0 +1,50 @@
+        # If no name, call `GetCollection()` on the configuration element, passing no name.
+        [String] $Name
+    )
+    Set-StrictMode -Version 'Latest'
```
Lint: newline between param block and first line of code.

splatteredbits commented on Carbon.IIS/Functions/Get-CIisCollection.ps1:
```diff
@@ -0,0 +1,50 @@
+        $sectionArgs["LocationPath"] = $LocationPath
+    }
+
+    $section = Get-CIisConfigurationSection @sectionArgs
```
Lint: only optional/conditional parameters should be splatted. `$section = Get-CIisConfigurationSection -SectionPath $SectionPath @getArgs`

splatteredbits commented on Carbon.IIS/Functions/Get-CIisCollection.ps1:
```diff
@@ -0,0 +1,50 @@
+
+    $section = Get-CIisConfigurationSection @sectionArgs
+
+    if ($Name)
```
Is there any way to validate that the collection name is correct? And that `$section` is a collection?

splatteredbits commented on Carbon.IIS/Functions/Get-CIisCollectionKeyName.ps1:
```diff
@@ -0,0 +1,29 @@
+function Get-CIisCollectionKeyName
+{
+    <#
```
Ditto docs.

splatteredbits commented on Carbon.IIS/Functions/Remove-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,68 @@
+# Fails if what you're removing doesn't exist.
```
Lint: comments about behavior should be in the documentation.

splatteredbits commented on Carbon.IIS/Functions/Remove-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,68 @@
+        [String] $CollectionName,
+
+        [Parameter(Mandatory)]
+        [String] $Value
```
IIS values aren't all strings. Change to `[Object]`.

splatteredbits commented on Carbon.IIS/Functions/Remove-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,68 @@
+        [Parameter(Mandatory)]
+        [String] $Value
+    )
+    Set-StrictMode -Version 'Latest'
```
Line: empty line between end of param block and first line of code.

splatteredbits commented on Carbon.IIS/Functions/Remove-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,68 @@
+        $collectionArgs['Name'] = $CollectionName
+    }
+
+    $collection = Get-CIisCollection @collectionArgs
```
Lint: only splat optional/conditional parameters: `$collection = Get-CIisCollection -SectionPath $SectionPath @collectionArgs`.

splatteredbits commented on Carbon.IIS/Functions/Remove-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,68 @@
+
+    $collection = Get-CIisCollection @collectionArgs
+
+    $uniqueKeyAttrName = Get-CIisCollectionKeyName -Collection $collection
```
`unique` and `key` mean the same thing, so shorten variable name to `keyAttrName`.

splatteredbits commented on Carbon.IIS/Functions/Remove-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,68 @@
+
+    $uniqueKeyAttrName = Get-CIisCollectionKeyName -Collection $collection
+
+    $itemToRemove =
```
Lint: only break a line across multiple lines if it is longer than 120 characters.

splatteredbits commented on Carbon.IIS/Functions/Remove-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,68 @@
+
+    if (-not $itemToRemove)
+    {
+        $msg = "Unable to find item with key ""$($Value)"" in ""$($LocationPath)/$($SectionPath)/$($CollectionName)" +
```
Sincre `$LocationPath` and `$CollectionName` are optional, make sure this message conditionally includes them so there aren't extra `/` in the message.

Use `:` between a location and the section path, e.g. `website/path:system.webServer/some/section` .

splatteredbits replied:
Remove redundant phrase "with key".

splatteredbits commented on Carbon.IIS/Functions/Remove-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,68 @@
+        return
+    }
+
+    $msg = "Removing item with ""$($uniqueKeyAttrName)"": ""$($Value)"" from """ +
```
Simplify `Removing item with ""$($uniqueKeyAttrName)"": ""$($Value)""` to `Removing item ""${Value}"" from ${target}.`

See previous notes about making sure the target part of the message doesn't contain extra `/` and uses a `:` to separate location path and section path.

splatteredbits commented on Carbon.IIS/Functions/Remove-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,68 @@
+    {
+        $msg = "Unable to find item with key ""$($Value)"" in ""$($LocationPath)/$($SectionPath)/$($CollectionName)" +
+               '".'
+        Write-Warning $msg
```
This should be an error. In order to let people ignore it, add `-ErrorAction $ErrorActionPreference`.

splatteredbits commented on Carbon.IIS/Functions/Remove-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,68 @@
+           "$($LocationPath)/$($SectionPath)/$($CollectionName)"""
+    Write-Information $msg
+    $collection.Remove($itemToRemove)
+    Save-CIisConfiguration
```
Output mesages should look like:

```text
system.webServer/path/to/collection
  - value
```

```text
location/path:section/path/collectionName
  - value
```

splatteredbits commented on Carbon.IIS/Functions/Set-CIisCollection.ps1:
```diff
@@ -0,0 +1,119 @@
+            </globalModules>
+        </system.webServer>
+    #>
+    [CmdletBinding()]
```
Does this actually work on PowerShell 5.1? I would expect a `DefaultParameterSetName='NotLocation'` for PowerShell to know what the parameter set is.

splatteredbits commented on Carbon.IIS/Functions/Set-CIisCollection.ps1:
```diff
@@ -0,0 +1,119 @@
+
+    process
+    {
+        Write-Verbose 'process'
```
Verbose messages are for users. Debug messages are for code authors. I believe these should be debug messages.

splatteredbits commented on Carbon.IIS/Functions/Set-CIisCollection.ps1:
```diff
@@ -0,0 +1,119 @@
+                $addItem.SetAttributeValue($key, $InputObject[0][$key])
+            }
+        }
+        else {
```
Lint: opening/closing braces should be on their own line.

splatteredbits commented on Carbon.IIS/Functions/Set-CIisCollection.ps1:
```diff
@@ -0,0 +1,119 @@
+    {
+        foreach ($collectionItem in $collection)
+        {
+            if ($collectionItem -notin $items -and $LocationPath)
```
Why the `-and $LocationPath` statement? That shouldn't matter?

CameronSWilliamson replied:
I was under the impression that I shouldn't modify anything in the default set so I was only clearing the collection *if* it was for a site. I should change where the `$save` variable is set though. It should be set as long as there are things being modified.

splatteredbits replied:
The purpose of `Set-CIisCollection` is to ensure a collection contains *only* the items passed in, regardless of site or global.

splatteredbits reviewed (changes requested):
I haven't looked at the tests yet. Once the code looks good, I'll review the tests.

splatteredbits commented on Carbon.IIS/Functions/Add-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,155 @@
+function Add-CIisCollectionItem
```
Rename this to `Set-CIisCollectionItem`.

CameronSWilliamson committed:
Added RunAsAdministrator requirement

CameronSWilliamson committed:
Cleaned up Add-CIisCollectionItem and renamed to Set-CIisCollectionItem

CameronSWilliamson committed:
Improved documentation and error handling for Get-CIisCollection

CameronSWilliamson committed:
Updated parameter documentation

CameronSWilliamson committed:
Updating documentation

CameronSWilliamson committed:
Cleaning up output messages and documentation

CameronSWilliamson committed:
Updated Remove-CIisCollectionItem tests to conform with new error messages

splatteredbits commented on CHANGELOG.md:
```diff
@@ -3,6 +3,24 @@
+
+### Added
+
+* Function `Add-CIisHttpHeader` for adding custom HTTP headers to IIS.
```
Rephrase to "Function `Add-CIisHttpHeader` for adding HTTP headers by adding them to the `system.webServer/httpProtocol/customHeaders` configuration collection."

splatteredbits commented on CHANGELOG.md:
```diff
@@ -3,6 +3,24 @@
+
+* Function `Add-CIisHttpHeader` for adding custom HTTP headers to IIS.
+* Function `Get-CIisCollection` for accessing items in a specific IIS collection.
+* Function `Get-CIisCollectionKey` for determining unique key for an IIS collection.
```
Typo? Should be "`Get-CIisCollectionKeyName`".

splatteredbits commented on Carbon.IIS/Carbon.IIS.psd1:
```diff
@@ -180,7 +186,7 @@
             # IconUri = ''
 
-            Prerelease = ''
+            Prerelease = 'rc1'
```
This should be empty. Build process will add a prerelease value if needed.

splatteredbits commented on Carbon.IIS/Carbon.IIS.psm1:
```diff
@@ -17,6 +17,7 @@ using namespace System.Management.Automation
 using namespace Microsoft.Web.Administration
 
 #Requires -Version 5.1
+#Requires -RunAsAdministrator
```
Why did you add this?

CameronSWilliamson replied:
We had discussed earlier that Carbon.IIS *has* to be run as administrator so I decided to add it as I was going through and fixing things. I can remove it if we actually don't need that requirement.

splatteredbits commented on Carbon.IIS/Functions/Add-CIisHttpHeader.ps1:
```diff
@@ -0,0 +1,70 @@
+{
+    <#
+    .SYNOPSIS
+    Adds a new header to the IIS configuration.
```
Rephrase "Adds a new HTTP header to IIS."

splatteredbits commented on Carbon.IIS/Functions/Add-CIisHttpHeader.ps1:
```diff
@@ -0,0 +1,70 @@
+    Adds a new header to the IIS configuration.
+
+    .DESCRIPTION
+    The `Add-CIisHttpHeader` function adds a new header to the IIS configuration. If adding a header for a specific
```
Make sure to give instructions on how to use the function, not just what it does. Rephrase to

```text
The `Add-CIisHttpHeader` function adds an HTTP header to IIS. Pass the header's name to the `Name` parameter
 and the header's value to the `Value` parameter. By default, the header is added to all HTTP responses. To add 
the header only to responses to a specific website, application, virtual directory, or directory, pass the path to 
that location to the `LocationPath` parameter.

The function adds the HTTP header by adding it to the `system.webServer/httPrototocl/customHeaders` 
configuration collection. 
```

splatteredbits commented on Carbon.IIS/Functions/Add-CIisHttpHeader.ps1:
```diff
@@ -0,0 +1,70 @@
+    Set-StrictMode -Version 'Latest'
+    Use-CallerPreference -Cmdlet $PSCmdlet -SessionState $ExecutionContext.SessionState
+
+    $addParameters = @{}
```
Rename to `setConditionalArgs`.

splatteredbits commented on Carbon.IIS/Functions/Get-CIisCollection.ps1:
```diff
@@ -0,0 +1,82 @@
+    Gets an instance of an IIS Collection
+
+    .DESCRIPTION
+    The `Get-CIisCollection` function gets the specified IIS collection. Pass the collection's IIS confuguration section
```
Typo "confuguration".

splatteredbits commented on Carbon.IIS/Functions/Get-CIisCollection.ps1:
```diff
@@ -0,0 +1,82 @@
+        return
+    }
+
+    if ($Name -and $collection.ElementTagName -ne $Name)
```
Why are you doing this check?

CameronSWilliamson replied:
It is a somewhat redundant check just to ensure that the returned collection name is in fact the same as the requested collection.

splatteredbits commented on Carbon.IIS/Functions/Get-CIisCollectionKeyName.ps1:
```diff
@@ -0,0 +1,28 @@
+    Returns the unique key for a configuration collection.
+
+    .DESCRIPTION
+    The `Get-CIisCollectionKeyName` locaties the mandatory attribute for an IIS configuration collection. This attribute
```
Typo: locaties.

splatteredbits commented on Carbon.IIS/Functions/Remove-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,83 @@
+{
+    <#
+    .SYNOPSIS
+    Removes a IIS Configuration element.
```
type: uppercase `Configuration`.

splatteredbits commented on Carbon.IIS/Functions/Remove-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,83 @@
+        $collectionArgs['Name'] = $CollectionName
+    }
+
+    $collection = Get-CIisCollection @collectionArgs -SectionPath $SectionPath
```
Make sure `$collection` has a value. If it doesn't, return. `Get-CIisCollection` should write an error if a collection doesn't exist.

```powershell
if ($null -eq $collection)
{
    return
}
```

splatteredbits commented on Carbon.IIS/Functions/Set-CIisCollection.ps1:
```diff
@@ -0,0 +1,118 @@
+{
+    <#
+    .SYNOPSIS
+    Changes the specied IIS Collection to only contain provided elements.
```
Rephrase: "Sets the exact contents of an IIS configuration section collection."

splatteredbits commented on Carbon.IIS/Functions/Set-CIisCollection.ps1:
```diff
@@ -0,0 +1,118 @@
+        Set-StrictMode -Version 'Latest'
+        Use-CallerPreference -Cmdlet $PSCmdlet -Session $ExecutionContext.SessionState
+        $items = [Collections.Generic.List[Microsoft.Web.Administration.ConfigurationElement]]::New()
+        $collection = Get-CIisCollection @PSBoundParameters
```
Handle if `$collection` doesn't exist.

splatteredbits commented on Carbon.IIS/Functions/Set-CIisCollection.ps1:
```diff
@@ -0,0 +1,118 @@
+        Use-CallerPreference -Cmdlet $PSCmdlet -Session $ExecutionContext.SessionState
+        $items = [Collections.Generic.List[Microsoft.Web.Administration.ConfigurationElement]]::New()
+        $collection = Get-CIisCollection @PSBoundParameters
+        $keyAttrName = Get-CIisCollectionKeyName -Collection $collection
```
Handle if `$keyAttrName` doesn't have a value.

splatteredbits commented on Carbon.IIS/Functions/Set-CIisCollection.ps1:
```diff
@@ -0,0 +1,118 @@
+
+    process
+    {
+        $addItem = $collection.CreateElement('add')
```
You should `foreach ($item in $InputObject)` .

splatteredbits commented on Carbon.IIS/Functions/Set-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,179 @@
+    .DESCRIPTION
+    The `Set-CIisCollectionItem` function adds a new item to an IIS configuration collection. Pass the collection's IIS
+    configuration section path to the `SectionPath` parameter and the value to add to the collection to the `Value`
+    parameter. The function adds that value to the collection if it doesn't already exist in the collection.
```
Add "or if it's value has changed."

splatteredbits commented on Carbon.IIS/Functions/Set-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,179 @@
+
+    [void]$msgs.Add($msgPrefix)
+
+    $collection = Get-CIisCollection @collectionArgs -SectionPath $SectionPath
```
What happens if the collection doesn't exist?

CameronSWilliamson committed:
Fixing Version number change

CameronSWilliamson committed:
Handling collections being null

CameronSWilliamson committed:
Fixing tests

CameronSWilliamson committed:
Fixing tests...again

CameronSWilliamson committed:
Fixing variable names

splatteredbits committed:
Updating dependencies.

splatteredbits committed:
Updating version.

splatteredbits committed:
Save private dependencies into a Modules directory.

splatteredbits committed:
Formatting

splatteredbits committed:
Creating Disable-CIisCollectionInheritance function.

splatteredbits committed:
Get-CIisCollection should work for configuration elements that don't have a section path (i.e. sites, app pools, etc.).

splatteredbits committed:
Get/Set/Remove collection functions should be able to operate on sites, application pools, virtual directories, and applications.

Added parameters to collection functions to pass in the configuration element objects on which to operate to support
using them with site, application pool, virtual directory, application, and other objects that aren't part of IIS's
configuration section model.

Discovered that the Microsoft.Web.Administration module doesn't properly save collection changes to site and
application pool objects. Instead of saving changes to the correct site/pool, changes were saved to the first site/pool
that didn't have the given collection. (See
https://stackoverflow.com/questions/77018655/microsoft-web-administration-not-saving-changes-to-a-sites-custom-log-fields-co.)
In researching a fix, found out that the IISAdministration module has Start-IISComitDelay and StopIISCommitDelay
functions. Taking inspiration from these, I created Suspend-CIisAutoCommit and Resume-CIisAutCommit functions to mimic
the IISAdministration module's funcationality and that fixed the issue.

splatteredbits committed:
Don't prefix property names with @ in output messages.

splatteredbits committed:
Creating function Get-CIisCollectionItem for getting collection items not the collection itself.

splatteredbits committed:
Remove-CIisCollectionItem should also operate directly on configuration elements.

splatteredbits committed:
Reduce code duplication. Set-CIisCollection should use Set-CIisCollectionItem and Remove-CIisCollectionItem to do the
actual changes.

splatteredbits committed:
Improving information messages for collection item updates.

splatteredbits committed:
Improving CHANGELOG.md.

splatteredbits committed:
Don't require run as administrator in case there's some situations where you can use part of Carbon.IIS without admin permissions.

decoyjoe commented on CHANGELOG.md:
```diff
@@ -3,6 +3,29 @@
+* Function `Get-CIisCollectionKeyName` for getting the unique key for an IIS configuration collection.
+* Function `Remove-CIisCollectionItem` for removing an item from an IIS configuration collection.
+* Function `Set-CIisCollection` for setting an IIS configuration collection.
+* Function `Set-CIisCollectionItem` for adding/update item in an IIS configuration collection.
```
Typo: `adding/update` should be `adding/updating`

decoyjoe commented on Carbon.IIS/Functions/Add-CIisHttpHeader.ps1:
```diff
@@ -0,0 +1,86 @@
+    .EXAMPLE
+    Add-CIisHttpHeader -LocationPath 'SITE_NAME' -Name 'X-AddHeader' -Value 'usingCarbon'
+
+    Demonstrates how to add a new HTTP header to the site `SITE_NAME`. After the above command runs, this will be in:
```
Typo: Remove the colon `:` from `in:`

decoyjoe commented on Carbon.IIS/Functions/Add-CIisHttpHeader.ps1:
```diff
@@ -0,0 +1,86 @@
+            <system.webServer>
+                <httpProtocol>
+                    <customHeaders>
+                        <add name="foo" value="bar" />
```
Typo: XML snippet doesn't match the example parameters.

decoyjoe commented on Carbon.IIS/Functions/Add-CIisHttpHeader.ps1:
```diff
@@ -0,0 +1,86 @@
+    #>
+    [CmdletBinding(DefaultParameterSetName='Global')]
+    param(
+        # The HTTP header name to add
```
Nit: Missing periods on parameter docs.

decoyjoe commented on Carbon.IIS/Functions/Add-CIisHttpHeader.ps1:
```diff
@@ -0,0 +1,86 @@
+        Set-StrictMode -Version 'Latest'
+        Use-CallerPreference -Cmdlet $PSCmdlet -SessionState $ExecutionContext.SessionState
+
+        $setConditionalArgs = @{}
```
Seems like an odd variable name. Perhaps `$locationParam` instead?

decoyjoe commented on Carbon.IIS/Functions/Add-CIisHttpHeader.ps1:
```diff
@@ -0,0 +1,86 @@
+    settings are updated). To add the header only to responses from a specific website, application, virtual directory,
+    or directory, pass the location's path to the `LocationPath` parameter.
+
+    The function adds the HTTP header to the `system.webServvver/httpProtocol/customHeaders` configuration collection.
```
Typo: `webServvver`

decoyjoe commented on Carbon.IIS/Functions/Disable-CIisCollectionInheritance.ps1:
```diff
@@ -0,0 +1,104 @@
+
+    Demonstrates how to disable inheritance on a collection under a site, directory, application, or virtual directory
+    by passing the location path to the site, directory, application, or vitual directory to the `LocationPath`
+    parameer.
```
Typo: `parameer`

decoyjoe commented on Carbon.IIS/Functions/Disable-CIisCollectionInheritance.ps1:
```diff
@@ -0,0 +1,104 @@
+    #>
+    [CmdletBinding()]
+    param(
+        [Parameter(Mandatory, ParameterSetName='ByConfigurationElement')]
```
Missing parameter doc.

decoyjoe commented on Carbon.IIS/Functions/Disable-CIisCollectionInheritance.ps1:
```diff
@@ -0,0 +1,104 @@
+    by passing the location path to the site, directory, application, or vitual directory to the `LocationPath`
+    parameer.
+    #>
+    [CmdletBinding()]
```
No default parameter set name?

decoyjoe commented on Carbon.IIS/Functions/Disable-CIisCollectionInheritance.ps1:
```diff
@@ -0,0 +1,104 @@
+
+    process
+    {
+        Set-StrictMode -Version 'Latest'
+        Use-CallerPreference -Cmdlet $PSCmdlet -SessionState $ExecutionContext.SessionState
```
Place these in a `begin` block?

splatteredbits replied:
Command does not accept pipeline input.

decoyjoe replied:
Then why does it use a `process {}` block? The function does not accept multiple items for any parameter.

decoyjoe commented on Carbon.IIS/Functions/Disable-CIisCollectionInheritance.ps1:
```diff
@@ -0,0 +1,104 @@
+        $displayPath = $collection.ElementTagName
+        if ($PSCmdlet.ParameterSetName -eq 'ByPath')
+        {
+            $displayPath = Get-CIisDisplayPath -SectionPath $sectionPath -LocationPath $LocationPath
```
Capitalize `$sectionPath` because it's a parameter.

decoyjoe commented on Carbon.IIS/Functions/Disable-CIisCollectionInheritance.ps1:
```diff
@@ -0,0 +1,104 @@
+
+        # The Microsoft.Web.Administration API does not expose any way of determining if a collection has a `clear`
+        # element, so we have to crack open the applicationHost.config file to look for it. :(
+        if (Test-CIisApplicationHostElement -XPath "${xpath}/clear" -LocationPath $LocationPath)
```
`$LocationPath` will be an empty string when using the `ByConfigurationElement` parameter set, is that ok?

splatteredbits replied:
Yes. The `Test-CIisApplicationHostElement` was written to handle an empty `LocationPath` argument.

decoyjoe commented on Carbon.IIS/Functions/Disable-CIisCollectionInheritance.ps1:
```diff
@@ -0,0 +1,104 @@
+        # element, so we have to crack open the applicationHost.config file to look for it. :(
+        if (Test-CIisApplicationHostElement -XPath "${xpath}/clear" -LocationPath $LocationPath)
+        {
+            return
```
Should we write a message if the the collection already contains clear, i.e. there is nothing to do?

decoyjoe commented on Carbon.IIS/Functions/Get-CIisCollection.ps1:
```diff
@@ -0,0 +1,96 @@
+{
+    <#
+    .SYNOPSIS
+    Gets an IIS configuration ollection.
```
Typo: `ollection`

splatteredbits replied:
Don't you mean "ypo"? 😆

decoyjoe commented on Carbon.IIS/Functions/Get-CIisCollection.ps1:
```diff
@@ -0,0 +1,96 @@
+    `Get-CIisCollectionItem`.
+
+    .EXAMPLE
+    $collection = Get-CIisCollection -LocationPath 'SITE_NAME' -SectionPath 'system.webServer/httpProtocol/' -Name 'customHeaders'
```
Trailing slash in `'system.webServer/httpProtocol/'` intentional?

decoyjoe commented on Carbon.IIS/Functions/Get-CIisCollection.ps1:
```diff
@@ -0,0 +1,96 @@
+    Demonstrates how to get the collection 'customHeaders' inside the section 'system.webServer/httpProtocol' for the
+    site 'SITE_NAME'.
+    #>
+    [CmdletBinding()]
```
Should there be a default parameter set name?

splatteredbits replied:
I think that's good practice, right? Always define a default parameter set if there's more than one? 🤔

decoyjoe replied:
I think so. Seems like good UX for the user to default to the easiest/preferred parameter set.

decoyjoe commented on Carbon.IIS/Functions/Get-CIisCollection.ps1:
```diff
@@ -0,0 +1,96 @@
+
+        # The configuration section path of the collection, or, if the configuration section is a parent of the
+        # collection, the configuration section path to the parent configuration section. If the configuration section
+        # is the parent of the collection, pass the collection name to the `CollectionName` parameter.
```
Typo: `CollectionName` should be `Name`

decoyjoe commented on Carbon.IIS/Functions/Get-CIisCollection.ps1:
```diff
@@ -0,0 +1,96 @@
+        $getArgs = @{}
+        if ($LocationPath)
+        {
+            $getArgs['LocationPath'] = $LocationPath
```
Rename to `$locationParam` ?

splatteredbits replied:
I prefer my variables that will be used to splat args to a function to be named after the function. That way, in the future, if/when I add other conditional args, I don't have to change the variable name.

decoyjoe commented on Carbon.IIS/Functions/Get-CIisCollection.ps1:
```diff
@@ -0,0 +1,96 @@
+            return
+        }
+
+        $displayPath = Get-CIisDisplayPath -SectionPath $SectionPath -LocationPath $LocationPath -SubSectionPath $Name
```
This should be using the hashtable to splat the `-LocationPath` parameter.

splatteredbits replied:
It could, but `Get-CIisDisplayPath` handles empty `LocationPath` argument value.

decoyjoe commented on Carbon.IIS/Functions/Get-CIisCollection.ps1:
```diff
@@ -0,0 +1,96 @@
+        return
+    }
+
+    return ,$collection
```
I would add a comment to mention that the preceding comma is explicitly necessary. It's not an obvious feature and looks like a typo to someone new to PowerShell.

decoyjoe commented on Carbon.IIS/Functions/Remove-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,150 @@
+    .DESCRIPTION
+    The `Remove-CIisCollectionItem` function removes an item from an IIS configuration collection. Pass the collection's
+    IIS configuration section path to the `SectionPath` parameter and the value to remove from the collection to the
+    `Value` parameter. This function removes that value from the collection if it exists.
```
Document what the function does if the value doesn't exist.

decoyjoe commented on Carbon.IIS/Functions/Remove-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,150 @@
+
+    Demonstrates how to remove the 'X-CarbonRemoveItem' header from the 'SITE_NAME' location.
+    #>
+    [CmdletBinding()]
```
Default parameter set?

decoyjoe commented on Carbon.IIS/Functions/Get-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,61 @@
+        # The `[Microsoft.Web.Administration.ConfigurationElement]` object whose collection items to get, or the parent
+        # element of the collection whose items to get. If this is the parent element, pass the name of the child
+        # element collection to the `CollectionName` parameter.
+        [Parameter(Mandatory, ParameterSetName='Direct')]
```
Use parameter set name `ByConfigurationElement` to be consistent with other functions.

decoyjoe commented on Carbon.IIS/Functions/Get-CIisCollection.ps1:
```diff
@@ -0,0 +1,96 @@
+        # The `[Microsoft.Web.Administration.ConfigurationElement]` object to get as a collection or the parent element
+        # of the collection element to get. If this is the parent element, pass the name of the child element collection
+        # to the `CollectionName` parameter.
+        [Parameter(Mandatory, ParameterSetName='Direct')]
```
Use parameter set name `ByConfigurationElement` to be consistent with other functions.

decoyjoe commented on Carbon.IIS/Functions/Remove-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,150 @@
+        [Parameter(Mandatory, ParameterSetName='ByConfigurationElement')]
+        [ConfigurationElement] $ConfigurationElement,
+
+        # The path to the section the item is located.
```
Missing a word, "where".

decoyjoe commented on Carbon.IIS/Functions/Set-CIisCollection.ps1:
```diff
@@ -0,0 +1,193 @@
+
+    .DESCRIPTION
+    The `Set-CIisCollection` function sets IIS configuration collection to a specific set of items. Pipe the collection
+    items to the function (or pass them to the `InputObject` parameter). You can pass just item vales if the collection
```
Typo: `vales`

decoyjoe commented on Carbon.IIS/Functions/Set-CIisCollection.ps1:
```diff
@@ -0,0 +1,193 @@
+    .DESCRIPTION
+    The `Set-CIisCollection` function sets IIS configuration collection to a specific set of items. Pipe the collection
+    items to the function (or pass them to the `InputObject` parameter). You can pass just item vales if the collection
+    items only have single values. Otherwise, pass a hashtable of name/value pairs.
```
I don't quite understand what this sentence means:

> You can pass just item values if the collection items only have single values.

decoyjoe commented on Carbon.IIS/Functions/Set-CIisCollection.ps1:
```diff
@@ -0,0 +1,193 @@
+    items to the function (or pass them to the `InputObject` parameter). You can pass just item vales if the collection
+    items only have single values. Otherwise, pass a hashtable of name/value pairs.
+
+    To make changes to a global configuration section, pass its path to the `SectionPath` parameter. To make changes to
+    a site, directory, application, or virtual directory, pass its pass location path to the `LocatinPath`. To make
```
Typo: `LocatinPath`

decoyjoe replied:
I would rephrase this because it sounds like you would use either `SectionPath` or `LocationPath` exclusively, when really you use `LocationPath` in addition to `SectionPath`.

decoyjoe commented on Carbon.IIS/Functions/Set-CIisCollection.ps1:
```diff
@@ -0,0 +1,193 @@
+
+    To make changes to a global configuration section, pass its path to the `SectionPath` parameter. To make changes to
+    a site, directory, application, or virtual directory, pass its pass location path to the `LocatinPath`. To make
+    changes to a specifci `[Microsoft.Web.Administration.ConfigurationElement]` object, pass it to the
```
Typo: `specifci`

decoyjoe commented on Carbon.IIS/Functions/Set-CIisCollection.ps1:
```diff
@@ -0,0 +1,193 @@
+            <system.webServer>
+                <defaultDocument>
+                    <files>
+                        <clear />
```
It's unclear why `<clear />` appears in that example.

decoyjoe commented on Carbon.IIS/Functions/Set-CIisCollection.ps1:
```diff
@@ -0,0 +1,193 @@
+        </location>
+
+    #>
+    [CmdletBinding()]
```
Set a default parameter set name?

decoyjoe commented on Carbon.IIS/Functions/Set-CIisCollection.ps1:
```diff
@@ -0,0 +1,193 @@
+        if (-not $keyAttrName)
+        {
+            $msg = "Failed to set IIS configuration collection ${displayPath} because it does not have a key attribute."
+            Write-Error -Message $msg
```
Add `-ErrorAction $ErrorActionPreference`

decoyjoe commented on Carbon.IIS/Functions/Set-CIisCollection.ps1:
```diff
@@ -0,0 +1,193 @@
+        Set-StrictMode -Version 'Latest'
+        Use-CallerPreference -Cmdlet $PSCmdlet -Session $ExecutionContext.SessionState
+
+        function Write-Message
```
This function isn't used anywhere.

decoyjoe commented on Carbon.IIS/Functions/Set-CIisCollection.ps1:
```diff
@@ -0,0 +1,193 @@
+            return
+        }
+
+        $itemsToRemove =
```
Write a message indicating what was removed. Unless the other functions already do that?

splatteredbits replied:
The other functions do that.

decoyjoe commented on Carbon.IIS/Functions/Remove-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,150 @@
+            return
+        }
+
+        if ($SkipCommit)
+        {
+            return $true
+        }
+
+        Save-CIisConfiguration
```
Should `Save-CIisConfiguration` be updated to check `$SkipCommit` itself? That way the other functions can simply always call `Save-CIisConfiguration` and not worry about whether to actually commit the changes or not?

splatteredbits replied:
This might be a future enhancement, but right now, only `Remove-CIisCollectionItem` and `Set-CIisCollectionItem` implement it. I'd have to write `Set-CIisSkipCommit` and `Reset-CIisSkipCommit` functions to manage it and then update `Set-CIisCollection` to call them both. Easier for now just to implement it as switches.

decoyjoe commented on Carbon.IIS/Functions/Set-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,261 @@
+
+    To configure a collection that is part of a global configuration section, pass the configuration section's path to
+    the `SectionPath` parameter. If the configuration section itself isn't a collection, pass the name of the collection
+    to the `Name` parameter. To configure a configuration section for a specific site, directory, application, or
```
Typo: `Name` should be `CollectionName`

decoyjoe commented on Carbon.IIS/Functions/Set-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,261 @@
+        [Parameter(Mandatory, ParameterSetName='Location')]
+        [String] $LocationPath,
+
+        # The path for the configuration section to edit
```
Nit: Missing period.

decoyjoe commented on Carbon.IIS/Functions/Set-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,261 @@
+    #>
+    [CmdletBinding(DefaultParameterSetName='Global')]
+    param(
+        [Parameter(Mandatory, ParameterSetName='Direct')]
```
Missing parameter doc.

decoyjoe commented on Carbon.IIS/Functions/Set-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,261 @@
+        [Parameter(Mandatory, ParameterSetName='Direct')]
+        [ConfigurationElement] $ConfigurationElement,
+
+        # The site to add the item to
```
Nit: Missing period.

decoyjoe commented on Carbon.IIS/Functions/Set-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,261 @@
+        if (-not $keyAttrName)
+        {
+            $msg = "Unable to find key for ${elementPath}."
+            Write-Error -Message $msg
```
Add `-ErrorAction $ErrorActionPreference`

decoyjoe commented on Carbon.IIS/Functions/Set-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,261 @@
+
+    process
+    {
+        if (-not $collection)
```
We'll never get here because this condition is already checked in the `begin` block.

splatteredbits replied:
PowerShell still executes `process` and `end` blocks when there's an error. If someone ignores/silently continues the error, the `process` and `end` blocks will also execute.

decoyjoe replied:
Oh, TIL.

decoyjoe commented on Tests/Suspend-CIisAutoCommit.Tests.ps1:
```diff
@@ -0,0 +1,69 @@
+        # [void]$customFields.Add('CLIENT-CERT-ISSUER', 'CLIENT-CERT-ISSUER', 'RequestHeader')
+        # [void]$customFields.Add('c-tp', 'c-tp', 'RequestHeader')
+        # [void]$customFields.Add('cert_header', 'cert_header', 'RequestHeader')
+        # [void]$customFields.Add('mycert', 'mycert', 'RequestHeader')
```
Are these all supposed to commented out?

splatteredbits committed:
Can't disable a generic configuration element collection inheritance without an xpath expression.

splatteredbits committed:
Only remove missing attributes from collection items when requested.

splatteredbits committed:
Code review feedback.

splatteredbits committed:
failing test

decoyjoe commented on CHANGELOG.md:
```diff
@@ -3,6 +3,29 @@
 
 # Carbon.IIS Changelog
 
+## 1.2.0
```
Add release date.

decoyjoe commented on Carbon.IIS/Functions/Disable-CIisCollectionInheritance.ps1:
```diff
@@ -0,0 +1,124 @@
+        [String] $Name
+    )
+
+    process
```
Using a `process {}` block but function doesn't accept pipeline input?

decoyjoe commented on Carbon.IIS/Functions/Set-CIisCollectionItem.ps1:
```diff
@@ -0,0 +1,282 @@
+
+
+    .EXAMPLE
+    Add-CIisCollectionItem -LocationPath 'SITE_NAME' -SectionPath 'system.webServer/defaultDocument' -CollectionName 'files' -Value 'hello.htm' -Attribute @{ 'name' = 'My File' }
```
Example shows `-Attribute` which isn't a valid parameter.

splatteredbits committed:
Code review feedback.
